src/Microsoft.TemplateEngine.Cli/New3Command.cs

```diff
@@ -281,37 +312,56 @@ private void HandlePostActions(TemplateCreationResult creationResult)
             postActionDispatcher.Process();
         }
 
-        private void DisplayTemplateList(bool showAll = false)
+        // Checks the result of TemplatesToDisplayInfoAbout()
+        // If they all have the same group identity, return them.
+        // Otherwise retun an empty list.
+        private IEnumerable<ITemplateInfo> TemplatesToShowDetailedHelpAbout()
         {
-            IReadOnlyCollection<ITemplateInfo> results = null;
+            IReadOnlyList<ITemplateInfo> candidateTemplates = TemplatesToDisplayInfoAbout();
 
-            try
+            if (candidateTemplates.Count() == 0)
             {
-                if (showAll)
-                {
-                    results = PerformAllTemplatesInContextQueryAsync().Result.Where(x => x.IsMatch).Select(x => x.Info).ToList();
+                return candidateTemplates;
+            }
 
-                    if(results.Count == 0)
-                    {
-                        results = PerformAllTemplatesInContextQueryAsync().Result.Select(x => x.Info).ToList();
-                    }
-                }
-                else
-                {
-                    results = _matchedTemplates.Where(x => x.IsMatch).Select(x => x.Info).ToList();
+            string firstGroupIdentity = candidateTemplates.First().GroupIdentity;
 
-                    if (results.Count == 0)
-                    {
-                        results = _matchedTemplates.Select(x => x.Info).ToList();
-                    }
-                }
+            if (candidateTemplates.Any(x => x.GroupIdentity != firstGroupIdentity))
+            {
+                return new List<ITemplateInfo>();
             }
-            catch (TemplateAuthoringException tae)
+
+            // they all have the same group identity
+            return candidateTemplates;
+        }
+
+        // If there are secondary matches, return them
+        // Else if there are primary matches, return them
+        // Otherwise return all templates in the current context
+        private IReadOnlyList<ITemplateInfo> TemplatesToDisplayInfoAbout()
+        {
+            IEnumerable<ITemplateInfo> templateList;
+
+            if (_matchedTemplatesWithSecondaryMatchInfo != null && _matchedTemplatesWithSecondaryMatchInfo.Count > 0)
             {
-                Reporter.Error.WriteLine(tae.Message.Bold().Red());
-                return;
+                templateList = _matchedTemplatesWithSecondaryMatchInfo.Select(x => x.Info);
+            }
+            else if (_matchedTemplates != null && _matchedTemplates.Where(x => x.IsMatch).Count() > 0)
+            {
+                templateList = _matchedTemplates.Where(x => x.IsMatch).Select(x => x.Info);
+            }
+            else
+            {
+                templateList = PerformAllTemplatesInContextQueryAsync().Result.Where(x => x.IsMatch).Select(x => x.Info);
```

Comment:
Discussed offline. Decided to make this synchronous since the template generators are no longer invoked during this flow.